### PR TITLE
Output DNS response data in JSON logs

### PR DESCRIPTION
**Pull request by edeca**
Add support for printing CNAME data in JSON output, using the same mechanism as TXT records.

### CODE REVIEW

**Comment by edeca** [comment]
A number of changes in 0516dc53495db2bfaa00ec7bf7f8ade93d06364b to reduce code duplication.  Also reduces the number of comparisons (e.g. ```if``` nested within ```select```).

I have tested these with A/PTR/MX/CNAME queries from pcap.

**Comment by edeca** [comment]
Added support for full parsing of the rcode header in DNS answer packets.  Where rcode isn't "no error" this is displayed in both DNS and JSON logs.

Note that this changes the current "No such domain" to "NXDOMAIN" in DNS logs.  This could be fixed if desired to maintain compatibility with anybody crazy enough to parse the DNS log.

When the rcode is not "no error" (for example ```NXDOMAIN``` or ```SERVFAIL```) it is unlikely that there will be answer RRs.  Therefore the rname from the query is used.

Because the rcode applies to a whole answer packet (not individual queries) it is impossible to determine which query RR caused the error.  Because of this most DNS servers currently reject multiple queries per packet.  Therefore each query RR is output instead with the relevant error code, likely to be ```FORMERR``` if ```queries > 1```.

File: src/app-layer-dns-tcp.c [hunk 1]
```diff
@@ -447,11 +447,15 @@ static int DNSReponseParseData(Flow *f, DNSState *dns_state, const uint8_t *inpu
         }
     }
 
-    /* see if this is a "no such name" error */
-    if (ntohs(dns_header->flags) & 0x0003) {
-        SCLogDebug("no such name");
+    /* parse rcode, e.g. "noerror" or "nxdomain" */
+    uint8_t rcode = ntohs(dns_header->flags) & 0x0F;
+    if (rcode <= 10 || (rcode >= 16 && rcode <= 22)) {
```
Comment (inliniac) [inline src/app-layer-dns-tcp.c]: Where do these numbers come from? Do they relate to DNS_RCODE_*? If so, please use those.

File: src/app-layer-dns-udp.c [hunk 2]
```diff
@@ -269,11 +269,15 @@ static int DNSUDPResponseParse(Flow *f, void *dstate,
         }
     }
 
-    /* see if this is a "no such name" error */
-    if (ntohs(dns_header->flags) & 0x0003) {
-        SCLogDebug("no such name");
+    /* parse rcode, e.g. "noerror" or "nxdomain" */
+    uint8_t rcode = ntohs(dns_header->flags) & 0x0F;
+    if (rcode <= 10 || (rcode >= 16 && rcode <= 22)) {
```
Comment (inliniac) [inline src/app-layer-dns-udp.c]: Where do these numbers come from? Do they relate to DNS_RCODE_*? If so, please use those.

File: src/output-json-dns.c [hunk 3]
```diff
@@ -157,7 +163,7 @@ static void OutputAnswer(LogDnsLogThread *aft, json_t *djs, DNSTransaction *tx,
             json_object_set_new(js, "rdata", json_string(a));
         } else if (entry->data_len == 0) {
             json_object_set_new(js, "rdata", json_string(""));
-        } else if (entry->type == DNS_RECORD_TYPE_TXT) {
+        } else if (entry->type == DNS_RECORD_TYPE_TXT || entry->type == DNS_RECORD_TYPE_CNAME || entry->type == DNS_RECORD_TYPE_MX || entry->type == DNS_RECORD_TYPE_PTR) {
```
Comment (inliniac) [inline src/output-json-dns.c]: Please break this line. We have a ~80 chars policy.

File: src/output-json-dns.c [hunk 4]
```diff
@@ -180,13 +186,55 @@ static void OutputAnswer(LogDnsLogThread *aft, json_t *djs, DNSTransaction *tx,
...
 static void LogAnswers(LogDnsLogThread *aft, json_t *js, DNSTransaction *tx, uint64_t tx_id)
 {
 
     SCLogDebug("got a DNS response and now logging !!");
 
-    if (tx->no_such_name) {
-        OutputAnswer(aft, js, tx, NULL);
+    /* rcode != noerror */
+    if (tx->rcode) {
+	/* Most DNS servers do not support multiple queries because
```
Comment (inliniac) [inline src/output-json-dns.c]: indent looks one off here

File: src/app-layer-dns-common.c [hunk 5]
```diff
@@ -783,7 +783,17 @@ const uint8_t *DNSReponseParse(DNSState *dns_state, const DNSHeader * const dns_
                 DNSStoreAnswerInState(dns_state, list, fqdn, fqdn_len,
                         ntohs(head->type), ntohs(head->class), ntohl(head->ttl),
                         data, 4, ntohs(dns_header->tx_id));
-            } else if (ntohs(head->type) == DNS_RECORD_TYPE_AAAA && ntohs(head->len) == 16) {
+            } else {
+                SCLogDebug("invalid length for A response data: %u", ntohs(head->len));
+		goto bad_data;
+	    }
```
Comment (inliniac) [inline src/app-layer-dns-common.c]: indent is one off

File: src/app-layer-dns-common.c [hunk 6]
```diff
@@ -759,22 +759,22 @@ const uint8_t *DNSReponseParse(DNSState *dns_state, const DNSHeader * const dns_
...
 
-            if (input + input_len < data + ntohs(head->len)) {
-                SCLogDebug("input buffer too small for data of len %u", ntohs(head->len));
-                goto insufficient_data;
-            }
-            SCLogDebug("TTL %u", ntohl(head->ttl));
+    SCLogDebug("head->len %u", ntohs(head->len));
+
+    if (input + input_len < data + ntohs(head->len)) {
+	SCLogDebug("input buffer too small for data of len %u", ntohs(head->len));
```
Comment (inliniac) [inline src/app-layer-dns-common.c]: indent is off

File: src/app-layer-dns-common.c [hunk 7]
```diff
@@ -791,46 +801,29 @@ const uint8_t *DNSReponseParse(DNSState *dns_state, const DNSHeader * const dns_
...
-                }
-
-                DNSStoreAnswerInState(dns_state, list, fqdn, fqdn_len,
-                        ntohs(head->type), ntohs(head->class), ntohl(head->ttl),
-                        cname, cname_len, ntohs(dns_header->tx_id));
-            }
+            } else {
+                SCLogDebug("invalid length for AAAA response data: %u", ntohs(head->len));
+                goto bad_data;
+	    }
```
Comment (inliniac) [inline src/app-layer-dns-common.c]: indent

**Comment by inliniac** [comment]
Please review my comments and submit a new branch and a new PR. Thanks!

**Pull request closed without merging**